File: services/product-service/src/main/java/com/Online_Auction/product_service/domain/Product.java
package com.Online_Auction.product_service.domain;

import java.time.LocalDateTime;
import java.util.List;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // ===== SELLER INFO =====
    private Long sellerId; // từ JWT

    // ===== BASIC INFORMATION =====
    private String name;

    @Column(columnDefinition = "TEXT")
    private String thumbnailUrl; // ảnh đại diện

    @ElementCollection
    @CollectionTable(name = "product_images", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "image_url")
    private List<String> images; // ảnh phụ (ít nhất 3 ảnh)

    @Column(columnDefinition = "TEXT")
    private String description; // mô tả (WYSIWYG)

    // ===== CATEGORY =====
    private Long parentCategoryId;
    private String parentCategoryName;

    private Long categoryId;
    private String categoryName;


    // ===== PRICING =====
    private Double startingPrice;

    @Builder.Default
    private Double currentPrice = null;
    private Double buyNowPrice;
    private Double stepPrice;     // bước giá

    // ===== AUCTION TIME =====
    private LocalDateTime createdAt;
    private LocalDateTime endAt;

    private boolean autoExtend; // có hỗ trợ tự động gia hạn không

    // ===== OTHER =====
    @Enumerated(EnumType.STRING)
    private ProductStatus status;

    @Builder.Default
    private Long bidCount = 0L;

    private Long currentBidder;

    public enum ProductStatus {
        ACTIVE,      // đang đấu giá
        FINISHED,    // đã kết thúc
        PENDING,     // đang chờ duyệt
        REJECTED
    }
}

File: services/category-service/internal/models/product.go
package models

import "time"

// Product represents a product in the auction system
type Product struct {
	tableName struct{} `pg:"products"`

	ID            int64     `json:"id" pg:"id,pk"`
	Name          string    `json:"name" pg:"name,notnull"`
	Description   string    `json:"description" pg:"description"`
	CategoryID    int64     `json:"category_id" pg:"category_id,notnull"`
	SellerID      int64     `json:"seller_id" pg:"seller_id,notnull"`
	StartingPrice float64   `json:"starting_price" pg:"starting_price,notnull"`
	CurrentPrice  float64   `json:"current_price" pg:"current_price"`
	BuyNowPrice   *float64  `json:"buy_now_price,omitempty" pg:"buy_now_price"`
	StepPrice     float64   `json:"step_price" pg:"step_price,notnull"`
	Status        string    `json:"status" pg:"status,notnull"`
	ThumbnailURL  string    `json:"thumbnail_url" pg:"thumbnail_url"`
	AutoExtend    bool      `json:"auto_extend" pg:"auto_extend,notnull,default:false"`
	EndAt         time.Time `json:"end_at" pg:"end_at,notnull"`
	CreatedAt     time.Time `json:"created_at" pg:"created_at,default:now()"`
	
	// Relations
	Category *Category       `json:"category,omitempty" pg:"rel:has-one"`
	Images   []*ProductImage `json:"images,omitempty" pg:"rel:has-many"`
}

// ProductImage represents product images
type ProductImage struct {
	tableName struct{} `pg:"product_images"`

	ProductID int64  `json:"product_id" pg:"product_id,notnull"`
	ImageURL  string `json:"image_url" pg:"image_url,notnull"`
}

// ProductListResponse represents paginated product list
type ProductListResponse struct {
	Products []*Product `json:"products"`
	Total    int        `json:"total"`
	Page     int        `json:"page"`
	PageSize int        `json:"page_size"`
	TotalPages int      `json:"total_pages"`
}

// ProductQueryParams represents query parameters for listing products
type ProductQueryParams struct {
	CategoryID int64  `query:"category_id"`
	Status     string `query:"status"`
	Page       int    `query:"page"`
	PageSize   int    `query:"page_size"`
	SortBy     string `query:"sort_by"`
	SortOrder  string `query:"sort_order"`
}
